adr/0002-masking-configuration-API.md

```diff
@@ -0,0 +1,199 @@
+# 2. API design for JSON masking configurations
+
+Date: 23-01-2024
+
+## Status
+
+Accepted by:
+
+- [Breus](https://github.com/Breus)
+- [gavlyukovskiy](https://github.com/gavlyukovskiy)
+- [donavdey](https://github.com/donavdey)
+
+## Context
+
+Before releasing version 1.x, we want to make the JSON masking configuration API as extensible as possible while
+providing predictable defaults. The current API was influenced by performance considerations which no longer apply due
+to certain optimizations done to the implementation. This resulted in a suboptimal API both in terms of extensibility
+and in terms of default behaviour.
+
+## Decisions
+
+### Change the default masking behaviors for all JSON types
+
+By default we
+1. We will not make a masking decision based on type. 
+    The main reason is that masking is explicitly requested by the user by specifying the target keys (either block
+    or allow) and it would be unexpected that we do not mask an explicitly requested key because it's a number and 
+    number masking is not enabled by default.
+2. We will enable length obfuscation by default as length on its own might leak information about the original value.
+3. We will mask all types into strings.
+    This is done to keep masking consistent across types and to make it apparent that the value was masked (while `"***"`
+    is a good indicator for strings, the `888` or `0` for numbers, or `false` for booleans is not)
+4. We will use different masks for each original type in to keep information about the original type, the strings will 
+    be masked as `"***"`, numbers as `"###"` and booleans as `"&&&"`.  
+    While this "leaks" additional information, that has only weak relation to original _data_, but more about 
+    implementation details and is more likely to be inferred by the key itself or be publically available.
+5. The `null` type (value) will remain unmasked until someone requests it, providing a good reason.
```

Comment:
Perhaps add that in this case it might be added as an **optional** feature (it won't affect the existing API)

Reply:
```suggestion
5. The `null` type (value) will remain unmasked, but we aim to make the API extendable to be able to support that if someone requests it, providing a good reason. 
```